scripts/parallel_pono.py

```diff
@@ -1,135 +1,133 @@
-    }
-
-    all_processes = []
-    queues = {}
-    name_map = {}
-
-    # this one gets updated on the fly as processes end
-    processes = []
-
-    for name, cmd in commands.items():
-        proc = subprocess.Popen(cmd, stdout=subprocess.PIPE)
-        processes.append(proc)
-        all_processes.append(proc)
-        name_map[proc] = name
-
-    def print_process_output(proc):
-        if proc.poll() is None:
-            proc.terminate()
-            proc.kill()
-        out, _ = proc.communicate()
-        print(out.decode('utf-8'))
-        print()
-        sys.stdout.flush()
-
-
-    shutdown = False
-    def handle_signal(signum, frame):
-        # send signal recieved to subprocesses
-        global shutdown
-        if not shutdown:
-            shutdown = True
-
-            global verbosity
-            if verbosity:
-                for proc in all_processes:
-                    print("{} output:".format(name_map[proc]))
-                    print_process_output(proc)
-                    print()
-                    sys.stdout.flush()
-            sys.exit(0)
+    def handle_signal(signum, frame) -> NoReturn:
+        # send signal received to subprocesses
+        for _, process, _ in processes.values():
+            process.send_signal(signum)
+        sys.exit(ReturnCode.UNKNOWN.value)
 
     signal.signal(signal.SIGINT, handle_signal)
     signal.signal(signal.SIGTERM, handle_signal)
 
-
-    while not shutdown:
-        for p in processes:
-            if p.poll() is not None:
-                # return code for unknown is 2
-                # anything higher than that is an error
-                # keep solving unless there are no more running processes
-                if p.returncode >= 2:
-                    processes.remove(p)
-                    # print unknown only if this is the last process
-                    if not processes:
-                        print_process_output(p)
-                        shutdown = True
-                else:
-                    # HACK don't return counter-examples from anything but bmc
-                    #      some others don't produce witnesses
-                    #      wouldn't expect K-Induction to ever win anyway
-                    if p.returncode == FALSE and name_map[p] != "BMC":
-                        processes.remove(p)
-                        # this shouldn't happen but let's handle it just in case
-                        if not processes:
-                            print_process_output(bmc)
-                            shutdown = True
-                            break
+    # Try current directory if pono is not on PATH.
+    executable = shutil.which("pono") or "./pono"
+
+    for name, options in ENGINE_OPTIONS.items():
+        cmd = [executable, "-k", str(args.bound), *options]
+        if args.witness_file:
+            cmd.extend(["--dump-btor2-witness", args.witness_file])
```

Comment:
To avoid race condition, it is safer to write to a unique temp file first.
We can later move the witness to the specified location.

Also, remember to clean up the temp files as requested by HWMCC.